core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -4004,6 +4488,99 @@ class KafkaApis(val requestChannel: RequestChannel,
+                                                  erroneous : mutable.Map[TopicIdPartition, ShareAcknowledgeResponseData.PartitionData],
+                                                ) : mutable.Map[TopicIdPartition, util.List[ShareAcknowledgementBatch]] = {
+
+    val acknowledgeBatchesMap = mutable.Map[TopicIdPartition, util.List[ShareAcknowledgementBatch]]()
+    shareFetchRequest.data().topics().forEach ( topic => {
+
+      if(!topicNames.asScala.contains(topic.topicId)) {
+        topic.partitions.forEach((partition: ShareFetchRequestData.FetchPartition) => {
+          val topicIdPartition = new TopicIdPartition(
+            topic.topicId,
+            new TopicPartition(null, partition.partitionIndex))
+          erroneous +=
+            topicIdPartition -> ShareAcknowledgeResponse.partitionResponse(topicIdPartition, Errors.UNKNOWN_TOPIC_ID)
+        })
+      }
+      else {
+        topic.partitions().forEach ( partition => {
+          val topicIdPartition = new TopicIdPartition(
+            topic.topicId(),
+            new TopicPartition(topicNames.get(topic.topicId()), partition.partitionIndex())
+          )
+          var exceptionThrown = false
+          val acknowledgeBatches = new util.ArrayList[ShareAcknowledgementBatch]()
+          breakable{
+            partition.acknowledgementBatches().forEach( batch => {
+              try {
+                acknowledgeBatches.add(new ShareAcknowledgementBatch(
+                  batch.firstOffset(),
+                  batch.lastOffset(),
+                  batch.acknowledgeTypes()
+                ))
+              } catch {
+                case e : IllegalArgumentException =>
+                  exceptionThrown = true
+                  erroneous += topicIdPartition -> ShareAcknowledgeResponse.partitionResponse(topicIdPartition, Errors.forException(e))
+                  break
+              }
+            })
+          }
+          if(!exceptionThrown && acknowledgeBatches.size() > 0) {
+            acknowledgeBatchesMap += topicIdPartition -> acknowledgeBatches
+          }
+        })
+      }
+    })
+    acknowledgeBatchesMap
+  }
+
+  def validateAcknowledgementBatches(
+                                      acknowledgementDataFromRequest : mutable.Map[TopicIdPartition, util.List[ShareAcknowledgementBatch]],
+                                      erroneous : mutable.Map[TopicIdPartition, ShareAcknowledgeResponseData.PartitionData]
+                                    ) : mutable.Set[TopicIdPartition] = {
+    val erroneousTopicIdPartitions: mutable.Set[TopicIdPartition] = mutable.Set.empty[TopicIdPartition]
+    acknowledgementDataFromRequest.foreach{ case (tp : TopicIdPartition, acknowledgeBatches : util.List[ShareAcknowledgementBatch]) =>
+      var prevEndOffset = -1L
+      breakable {
+        acknowledgeBatches.forEach(batch => {
+          if (batch.firstOffset > batch.lastOffset) {
```

Comment:
can we not club all of this "if" and the below "if" conditions into a single if?

Reply:
Thanks for the review ! Yes we can but I did that for better readability, without affecting the execution at all.

Reply:
Yeah it's much readable this way, I agree.